Code/autopkglib/URLDownloader.py

```diff
@@ -116,45 +104,99 @@ class URLDownloader(Processor):
         },
     }
 
-    def main(self):
-        if not is_mac():
-            raise ProcessorError("This processor is Mac-only!")
-        # clear any pre-exising summary result
+    def getxattr(self, attr):
+        """Get a named xattr from a file. Return None if not present"""
+        if attr in xattr.listxattr(self.env["pathname"]):
+            return xattr.getxattr(self.env["pathname"], attr)
+        return None
+
+    def prepare_curl_cmd(self, pathname_temporary):
+        """Assemble curl command and return it."""
+        curl_cmd = [
+            super(URLDownloader, self).curl_binary(),
+            "--silent",
+            "--show-error",
+            "--no-buffer",
+            "--fail",
+            "--dump-header",
+            "-",
+            "--speed-time",
+            "30",
+            "--location",
+            "--url",
+            self.env["url"],
+            "--output",
+            pathname_temporary,
+        ]
+
+        super(URLDownloader, self).add_curl_common_opts(curl_cmd)
+
+        # if file already exists and the size is 0, discard it and download again
+        if (
+            os.path.exists(self.env["pathname"])
+            and os.path.getsize(self.env["pathname"]) == 0
+        ):
+            os.remove(self.env["pathname"])
+
+        # if file already exists, add some headers to the request
+        # so we don't retrieve the content if it hasn't changed
+        if os.path.exists(self.env["pathname"]):
+            self.existing_file_size = os.path.getsize(self.env["pathname"])
+            etag = self.getxattr(self.xattr_etag)
+            last_modified = self.getxattr(self.xattr_last_modified)
+            if etag:
+                curl_cmd.extend(["--header", "If-None-Match: %s" % etag])
+            if last_modified:
+                curl_cmd.extend(["--header", "If-Modified-Since: %s" % last_modified])
+
+        return curl_cmd
+
+    def clear_vars(self):
```

Comment:
`clear_vars()` initializes the variables. This could be perhaps moved into class constructor. Split from `main()`: [lines](https://github.com/autopkg/autopkg/blob/15c6307254d4ef1d1dbb8a2648a204a340f5b967/Code/autopkglib/URLDownloader.py#L123) 123..128